transform/transformers/MWSSTransformer.py

```diff
@@ -0,0 +1,437 @@
+            Survey.parse_timestamp(data.get("submitted_at", ts.isoformat())),
+            data.get("metadata", {}).get("user_id"),
+            ''.join(i for i in ruRef if i.isdigit()),
+            ruRef[-1] if ruRef and ruRef[-1].isalpha() else "",
+            data.get("collection", {}).get("period")
+        )
+        if any(i is None for i in rv):
+            log.warning("Missing an id from {0}".format(rv))
+            return None
+        else:
+            return rv
+
+
+class Processor:
+    """
+    Principles for processors:
+
+    * method is responsible for range check according to own logic.
+    * parametrisation is possible; use functools.partial
+    * returns data of the same type as the supplied default.
+
+    """
+
+    @staticmethod
+    def aggregate(qId, data, default, *args, weights=[], **kwargs):
+        """
+        Calculates the weighted sum of a question group.
+
+        """
+        try:
+            return type(default)(
+                Decimal(data.get(qId, 0)) +
+                sum(Decimal(scale) * Decimal(data.get(q, 0)) for q, scale in weights)
+            )
+        except (InvalidOperation, ValueError):
+            return default
+
+    @staticmethod
+    def evaluate(qId, data, default, *args, group=[], convert=bool, op=operator.or_, **kwargs):
+        """
+        Performs a map/reduce evaluation of a question group.
+
+        """
+        try:
+            groupVals = [data.get(qId, None)] + [data.get(q, None) for q in group]
+            data = [convert(i) for i in groupVals if i is not None]
+            return type(default)(reduce(op, data))
+        except (TypeError, ValueError):
+            return default
+
+    @staticmethod
+    def mean(qId, data, default, *args, group=[], **kwargs):
+        """
+        This function calculates the value of a field based on the mean
+        of all values in a declared question group.
+
+        """
+        try:
+            groupVals = [data.get(qId, None)] + [data.get(q, None) for q in group]
+            data = [Decimal(i) for i in groupVals if i is not None]
+            divisor = len(data) or 1
+            rv = sum(data) / divisor
+            return type(default)(rv)
+        except (AttributeError, InvalidOperation, TypeError, ValueError):
+            return default
+
+    @staticmethod
+    def events(qId, data, default, *args, group=[], **kwargs):
+        """
+        This function returns a sequence of events in time from values in a
+        declared question group.
+
+        """
+        try:
+            groupVals = [data.get(qId, None)] + [data.get(q, None) for q in group]
+            data = sorted(filter(
+                None, (Survey.parse_timestamp(i) for i in groupVals if i is not None)
+            ))
+            if all(isinstance(i, type(default)) for i in data):
+                return data
+            else:
+                return type(default)(data)
+        except (AttributeError, TypeError, ValueError):
```

Comment:
Again, not really an error condition, more part of the logic as to whether to return the default value.